### PR TITLE
Added TestDox support for legible, requirements based PyTest output

### PR DESCRIPTION
**Pull request by cartertinney**
*Added pytest-testdox to our dev environment and configured it to run automatically
*Refactored **some** tests to fix up the worst problems with test output
*Added a few missing tests

What is pytest-testdox? It's a pytest output formatter that formats things... essentially like a requirements doc. It's not perfect, but it's a lot closer than we were. Give this branch a pull, and try running pytest, see what you think.

Note that I have still not standardized any kind of format for the phrasing or organization of the requirements. I just fixed the most egregious tests up so there's a mostly legible output.

### CODE REVIEW

**Comment by olivakar** [comment]
```
azure-iot-common/tests/test_asyncio_compat.py                                                                                                                                             
Get Running Loop
 [x] returns currently running event loop
 [x] raises runtime error if no running event loop
 >>> returns currently running event loop py36orless compat
 >>> raises runtime error if no running event loop py36orless compat
```

There are 4 tests in this class. The last 2 tests' output gets shown in a different format than the first 2 tests. Do you know why is that the case ? I think it is because of the word `py36orless`. They want it to be more readable (english language wise) I guess

**Comment by olivakar** [comment]
```python
azure-iot-hub-devicesdk/tests/test_sync_clients.py                                                                                                                                             
Module Client
 [x] from authentication provider instantiates client[SymmetricKey-mqtt]
 >>> from authentication provider instantiates client[SymmetricKey-amqp]
 >>> from authentication provider instantiates client[SymmetricKey-http]
```
In this parametrized test I am guessing since AMQP and HTTP are not implemented hence the output is different. Is that true or is it something else ?

**Comment by olivakar** [comment]
```python
 [x] from authentication provider boundary case transport name[ALL CAPS]
 [x] from authentication provider boundary case transport name[mIxEd CaSe]
 [x] from authentication provider bad input raises error transport name[]
```
In this scenario however . the last one seems like the bad input is actually an empty input. but that is not the case. The input itself is the word `bad input` right ?  may be its best if the `ids=[""]` is not there at the top coz this is not a multi param test. You are passing `SymmetricKey` as the fixed param and `bad input` as the transport name right ?

File: azure-iot-common/tests/test_asyncio_compat.py [hunk 1]
```diff
@@ -14,73 +14,75 @@ def dummy_coroutine():
...
+        assert spy_get_running_loop.call_args == mocker.call()
+
+    @pytest.mark.skipif(sys.version_info < (3, 7), reason="Requires Python 3.7+")
+    async def test_raises_runtime_error_if_no_running_event_loop(self, mocker):
+        mocker.patch.object(asyncio, "get_running_loop", side_effect=RuntimeError)
+        with pytest.raises(RuntimeError):
+            asyncio_compat.get_running_loop()
+
+    @pytest.mark.skipif(sys.version_info >= (3, 7), reason="Requires Python 3.6 or below")
+    async def test_returns_currently_running_event_loop_py36orless_compat(self, mocker, event_loop):
```
[the remark this answers is not in the document]
Comment (cartertinney) [reply azure-iot-common/tests/test_asyncio_compat.py]: This was intentional. It's hard to read if it's not one word. 

"returns currently running event loop py36 or less compat"

It's not clear what compat is referring too. Returns currently running event loop py36. Or less compat?

**Pull request merged**